userguide/tutorials/payment_plugin.adoc

```diff
@@ -1,61 +1,575 @@
+public PaymentTransactionInfoPlugin purchasePayment(UUID kbAccountId, UUID kbPaymentId, UUID kbTransactionId,
+	UUID kbPaymentMethodId, BigDecimal amount, Currency currency, Iterable<PluginProperty> properties,
+	CallContext context) throws PaymentPluginApiException {
+	logger.info("purchasePayment, kbAccountId=" + kbAccountId);
+	PaymentTransactionInfoPlugin paymentTransactionInfoPlugin; 
+	String mandate = getMandateId(kbAccountId, context);  //retrieve mandateId from Kill Bill tables
+	logger.info("MandateId:", mandate);
+	if (mandate != null) {
+		logger.info("Processing payment");
+		try {
+			String idempotencyKey = PluginProperties.findPluginPropertyValue("idempotencykey", properties); 
+			com.gocardless.services.PaymentService.PaymentCreateRequest.Currency goCardlessCurrency = convertKillBillCurrencyToGoCardlessCurrency(currency);
+			Payment payment = client.payments().create().withAmount(Math.toIntExact(KillBillMoney.toMinorUnits(currency.toString(), amount)))
+					.withCurrency(goCardlessCurrency).withLinksMandate(mandate).withIdempotencyKey(idempotencyKey)
+					.execute();
+			List<PluginProperty> outputProperties = new ArrayList<PluginProperty>();
+			outputProperties.add(new PluginProperty("paymentId", payment.getId(), false));
+			paymentTransactionInfoPlugin = new GoCardlessPaymentTransactionInfoPlugin(
+					kbPaymentId, kbTransactionId, TransactionType.PURCHASE, amount, currency, PaymentPluginStatus.PROCESSED, null,
+					null, String.valueOf(payment.getId()), null, new DateTime(), new DateTime(payment.getCreatedAt()), outputProperties);
+			logger.info("Payment processed, PaymentId:", payment.getId());
+		} catch (GoCardlessApiException e) {
+			paymentTransactionInfoPlugin = new GoCardlessPaymentTransactionInfoPlugin(
+					kbPaymentId, kbTransactionId, TransactionType.PURCHASE, amount, currency,  PaymentPluginStatus.ERROR, e.getErrorMessage(),				String.valueOf(e.getCode()), null, null, new DateTime(), null, null);
+			logger.warn("Error occured in purchasePayment", e.getType(), e);
+		}
+	}
+	else {
+		logger.warn("Unable to fetch mandate, so cannot process payment");
+		paymentTransactionInfoPlugin = new GoCardlessPaymentTransactionInfoPlugin(
+				kbPaymentId, kbTransactionId, TransactionType.PURCHASE, amount, currency, PaymentPluginStatus.CANCELED, null, null, null, null, new DateTime(), null, null);
+	}
+	return paymentTransactionInfoPlugin;
+}
+
+* Like the other methods, the `purchasePayment` method accepts  `Iterable<PluginProperty> properties` as a parameter. 
+
+* In addition, it also accepts parameters corresponding to `amount` (specifies the amount to charge the customer) and `currency` (specifies the currency to be used)
+
+* If you recall, the `addPaymentMethod` stores the *mandate id* in the Kill Bill database. This is first retrieved and assigned to `mandate` 
+
+* Next, the `idempotencyKey` is retrieved from the `properties` passed in. The  `idempotencyKey` is a GoCardless specific value.
```

Comment:
Per the GoCardless documentation, their API will ensure this payment is only ever created once per `idempotencyKey`. So here, we could specify `kbPaymentId` as the `idempotencyKey`, to ensure at most a single payment per `kbPaymentId` is created.